### PR TITLE
Access Credentials API (with dynamic propagation)

### PR DESCRIPTION
**Pull request by davidvossel**
# Overview 

This PR provides a dedicated API for supplying SSH public keys and guest user passwords in a way that is both dynamic (can be applied to running VMs) and decoupled from the cloud-init user data. The Idea here is that users can define their ssh public keys  and user passwords in secrets, and assign those to a VMI. Using the qemu guest agent, can watch those secrets in the VMI's pod using the downward API, and dynamically inject changes to credentials into a running VM.

This is a follow up to the abandoned PR #3930

# User Stories 

**Granting Shared SSH Provisioning access used across all VM infrastructure in namespace**
- An ops team can create a single secret containing all the authorized ssh public keys their ansible provisioner uses.
- Users can create VMs and attach the standard authorization keys used to automatically provision the VMs

**Granting team member shared ssh access to VMs in namespace**
- All team member's public ssh keys are added to a k8s secret
- All VMs reference this secret as their ssh public key access credential

**Dynamic Infrastructure Key rotation**
- An ops team can create a single secret containing all the authorized ssh public keys their ansible provisioner uses.
- VMs created in namespace all reference this secret which injects ssh public keys. 
- Ops team can rotate keys by updating new public keys in secret which dynamically propagates to each running guest.

**Dynamically Granting/Revoking user access**
- All team member's public ssh keys are added to a k8s secret
- All VMs reference this secret as their ssh public key access credential
- Adding or revoking access to VMs would be as simple as adding/removing a key from the k8s secret. This would propagate to all VMI pods and ultimately into the guest. 

# AccessCredentials API

The Access Credentials API is structured in a way that should make it easy for us to expand in the future. Each credential added to a VMI will have a specific type (such as SSH or Password), and each type has a source and propagation method.

So, you pick what type of credential it is you want to inject, then you pick where the credential is coming from (such as a secret) and how it is injected (such as using the qemu guest agent).  

Organizing the API in this way lets us expand the api in the future to new credential types (like x509 perhaps) and new sources and methods for propagating the credential into the guest. 

## SSH Key Example
create a secret
```
kubectl create secret generic my-pub-key --from-file=id_rsa.pub
```

Assign that secret to a VMI using the AccessCredential API
```
apiVersion: kubevirt.io/v1alpha3
kind: VirtualMachineInstance
metadata:
  name: agent
spec:
  domain:
    devices:
      disks:
      - disk:
          bus: virtio
        name: disk0
    resources:
      requests:
        cpu: 100m
        memory: 1G
  accessCredentials:
  - sshPublicKey:
      source:
        secret:
          secretName: my-pub-key
      propagationMethod:
        qemuGuestAgent:
          users:
          - "fedora"
  volumes:
  - containerDisk:
      image: registry:5000/kubevirt/fedora-cloud-container-disk-demo:devel
      imagePullPolicy: IfNotPresent
    name: disk0
```

## User Passwords Example

Create a secret where the user is the key and the password is the value
```
echo -n 'some-password' > ./password.txt
kubectl create secret generic my-fedora-password --from-file=fedora=./password.txt
```

```
apiVersion: kubevirt.io/v1alpha3
kind: VirtualMachineInstance
metadata:
  name: agent
spec:
  domain:
    devices:
      disks:
      - disk:
          bus: virtio
        name: disk0
    resources:
      requests:
        cpu: 100m
        memory: 1G
  accessCredentials:
  - userPassword:
      source:
        secret:
          secretName: my-pub-key2
      propagationMethod:
        qemuGuestAgent: {}
  volumes:
  - containerDisk:
      image: registry:5000/kubevirt/fedora-cloud-container-disk-demo:devel
      imagePullPolicy: IfNotPresent
    name: disk0
```

**Release note**:
```release-note
AccessCredentials API for dynamic user/password and ssh public key injection
```

### CODE REVIEW

**Comment by fabiand** [comment]
Neat work.

Questions on API:

```
      propagationMethod:
        qemuGuestAgent: {}
```

I suppose proagation method is optional?

same question for
```
      propagationMethod:
        qemuGuestAgent:
          authorizedKeysFiles:
          - filePath: "/home/fedora/.ssh/authorized_keys"
```
Or in this case, could `~/.ssh/authorized_keys` work as well? (To have a sane default)

**Comment by davidvossel** [comment]
>Questions on API:
>>    propagationMethod:
>>        qemuGuestAgent: {}
>I suppose proagation method is optional?

right now, propagation method is required. There are two types, qemuAgent and configDrive.

>same question for
>>      propagationMethod:
>>        qemuGuestAgent:
>>          authorizedKeysFiles:
>>          - filePath: "/home/fedora/.ssh/authorized_keys"
>Or in this case, could ~/.ssh/authorized_keys work as well? (To have a sane default)

@fabiand I went back and forth on this one quite a bit. I think the only sane default is /root/.ssh/authorized_keys. I'd expect that to be the same as `~/.ssh/authorized_keys` since the agent runs as root. The difference is that we're not leaving room for ambiguity around exactly what is expected if we hard code to /root/.ssh/authorized_keys as default. 

That said. I feel a little uneasy implicitly exposing root by default. This is why I left this path list as being user defined.

**Comment by davidvossel** [comment]
I've finished all my `TODOs` at this point. Unit and functional tests are in place.The PR should be ready for review

**Comment by davidvossel** [comment]
I'm converting this back to a draft for a bit as I modify the API to align more closely with the work the qemu project is introducing for their ssh api (https://bugzilla.redhat.com/show_bug.cgi?id=1885332)

**Comment by davidvossel** [comment]
PR is updated to reflect how (https://bugzilla.redhat.com/show_bug.cgi?id=1885332) will work. I also added conditions to indicate on the VMI status when the credentials sync or if there are issues during the sync.

The biggest change here is we now set ssh keys by the internal guest user and don't require specifying a authorized_keys file.

File: pkg/virt-controller/services/template.go
Comment (rmohr) [inline pkg/virt-controller/services/template.go:626]: I see that we use this pattern all over the place. Probably not the best best solution. Seems like we can end up with collisions. Since we never generate volumes for users if they are not in the `volumes` list of the VMI, I wonder why we ever started with adding suffixes. :/

File: pkg/virt-handler/vm.go
Comment (rmohr) [inline pkg/virt-handler/vm.go:660]: Can we also have events sent? Especially of something goes wrong?

Comment (davidvossel) [reply pkg/virt-handler/vm.go:660]: yup, good idea.

File: pkg/virt-controller/services/template.go
Comment (davidvossel) [reply pkg/virt-controller/services/template.go:626]: In this case, the access credential secret name is not a volume, so the suffix is actually meant to help reduce collisions. Using the secret name + suffix allows us to easily infer the secret's direction from within the pod. 

any thoughts on what might be a potentially better solution?

File: pkg/virt-handler/vm.go
Comment (rmohr) [reply pkg/virt-handler/vm.go:660]: Great. Let's ensure that people can see without looking into logs that they get propagated or when it fails that they see in the events why it failed.

Comment (davidvossel) [reply pkg/virt-handler/vm.go:660]: done

**Comment by rmohr** [comment]
/approve
/hold

Design and API looks good to me.
Needs a proper code review from someone else.

**Comment by vladikr** [comment]
I've already looked at this PR and didn't find anything critical. Everything looked good to me.
I'll go over the code again in a bit.

File: pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go
Comment (vladikr) [inline pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go:536]: I wonder if we should distinguish between different types of agent errors and fail early (or alert the user) if the agent is not installed / not available.
Something like we do here: https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-launcher/virtwrap/manager.go#L554

**Review by vladikr — COMMENTED** [review]
It all looks great to me! It's a huge work!

I'm only wondering if we alert the user enough in the case when the guest agent is not installed or not available/connected or stuck...

File: pkg/virt-launcher/virtwrap/manager.go
Comment (davidvossel) [reply pkg/virt-launcher/virtwrap/manager.go:1070]: oh nice catch! yes i need to handle this errror. I think this function origially didn't return an error so i misssed handling it once things changed.

File: pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go
Comment (davidvossel) [reply pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go:536]: take a look at what I'm doing with the ```l.reportAccessCredentialResult``` function in ```pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go```

What happens is whenever an error occurs, that error typically indicates both where it occurred (like during file write) with the libvirt error message appended to the end. That error then gets placed on the domain's metadata which is read by virt-handler. Virt-handler then adds both a condition and reports an event to indicated to the user whether or not access credential syncing succeeded or not (with a message indicating why it didn't succeed)

Maybe i could interpret the libvirt error code, but from what I've seen so far it seems like the libvirt error message itself contains a human readable string that indicates the issue already.

File: pkg/virt-launcher/virtwrap/manager.go
Comment (davidvossel) [reply pkg/virt-launcher/virtwrap/manager.go]: fixed!

**Comment by davidvossel** [comment]
> I'm only wondering if we alert the user enough in the case when the guest agent is not installed or not available/connected or stuck...

the user gets a condition and an event on the VMI when thsi occurs. example https://github.com/kubevirt/kubevirt/blob/4dfb7b1471d1f4023abfa9d65fd1404b7cf7840c/pkg/virt-handler/vm.go#L677

File: pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go
Comment (vladikr) [reply pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go:536]: Thanks for the explanation! 
Yeah, I think it's good.

What I meant with my comment was that we could stop watching the directories when we determine that the agent is not even installed.
But I think that this won't give the user a chance to install it after seeing the event.

**Comment by vladikr** [comment]
Thanks!
/lgtm

**Comment by davidvossel** [comment]
/hold

need to fix some unrelated flaky test related to controller eviction before this goes in. Something about the timing of the new tests in this pr is causing that eviction test to fail more frequently.

**Review by Donkey23 — APPROVED** [review]

**Review by Donkey23 — APPROVED** [review]

**Review by Donkey23 — APPROVED** [review]

**Comment by yichengliu58** [comment]
hi @rmohr this is an amazing feature! I just want to ask are there any release plans for this feature ? Like in v0.36.0 will this be added into ?

**Comment by rmohr** [comment]
Yes, this feature from @davidvossel will be available in v0.36.

**Comment by yichengliu58** [comment]
@davidvossel Hi I just tried this access credentials feature but failed to create the VM. I used the VM definition yaml like this:
```yaml
apiVersion: kubevirt.io/v1alpha3
kind: VirtualMachine
metadata:
  labels:
    kubevirt.io/vm: test-vm-2
  name: test-vm-2
  namespace: demo
spec:
  runStrategy: RerunOnFailure
  template:
    metadata:
      labels:
        kubevirt.io/vm: test-vm-2
    spec:
      terminationGracePeriodSeconds: 10
      accessCredentials:
      - userPassword:
          source:
            secret:
              secretName: my-pub-key2
          propagationMethod:
            qemuGuestAgent: {}
      domain:
      # ...
```

and use
`kubectl apply -f vm.yaml`

the output:
`The request is invalid: : spec.template.spec.accessCredentials in body is a forbidden property`

I didn't see any syntax errors in my yaml, not sure why this can't be applied...

**Pull request merged**
